software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,252 @@
+      ['.sdata'            , ['.srodata.cst16','.srodata.cst8','.srodata.cst4',
+                              '.srodata.cst2','.srodata*','.sdata','.sdata.*',
+                              '.gnu.linkonce.s.*']],
+      ['.sbss'             , ['.sbss','.sbss.*','.gnu.linkonce.sb.*','.scommon']],
+      ['.bss'              , ['.bss','.bss*']],
+      ['.tdata'            , ['.tdata','.tdata*']],
+      ['.tbss'             , ['.tbss','.tbss*']],
+      ['.eh_frame'         , ['.eh_frame','.eh_frame*']],
+      ['.striped.data.dmem', ['.striped.data']],
+      ['.rodata.dram'      , ['.rodata','.rodata*']],
+      ['.dram'             , ['.dram','.dram.*', '.*']],
+      ]
+
+    sections = "SECTIONS {\n\n"
+
+    # DMEM sections
+    for i, m in enumerate(section_map):
+      sec = m[0]
+      laddr = '0x1000'
+      in_sections = m[1]
+      in_objects = '*' if self._default_data_loc == 'dmem' \
+          else '*bsg_manycore_lib.a:'
+
+      if re.search(".dram$", sec) != None:
+        continue
+
+      if re.search(".dmem$", sec):
+        in_objects = '*'
+      else:
+        sec += '.dmem'
+      
+      if sec != ".dmem":
+        prev_sec = section_map[i-1][0]
+
+        if re.search(".dmem$", prev_sec) == None:
+          prev_sec += '.dmem'
+
+        laddr = "LOADADDR({0}) + ADDR({1}) - ADDR({0})".format(
+            prev_sec, sec)
+
+      sections += self._section(sec, laddr, 'DMEM_VMA', None,
+          in_sections, in_objects)
+
+    # DRAM sections
+    for i,m in enumerate(section_map):
+      sec = m[0]
+      vaddr = ""
+      in_sections = m[1]
+      in_objects = '*'
+
+      if re.search(".dmem$", sec) != None:
+        continue
+
+      if sec == ".text.dram":
+        vaddr = "0x0"
+
+      if re.search(".dram$", sec) == None and self._default_data_loc == 'dram':
+        sec += '.dram'
+
+      if self._default_data_loc == 'dram' or re.search(".dram$", sec) != None:
+        sections += self._section(sec, vaddr, None, 'DRAM_LMA',
+            in_sections, in_objects)
+
+      if sec == '.text.dram':
+        sections += ". = . + 0x80000000;\n\n"
+
+    # Symbols
+    if self._default_data_loc == 'dmem':
+      sections += "_gp = ADDR(.sdata.dmem) + 0x800;\n"
+    else:
+      sections += "_gp = ADDR(.sdata.dram) + 0x800;\n"
+    sections += "_sp = 0x{0:08x};\n".format(self._sp)
+    sections += "_bsg_data_start_addr = 0x{0:08x};\n".format(_DMEM_VMA_START)
+    sections += "_bsg_data_end_addr = ADDR(.striped.data.dmem) + " \
+                "SIZEOF(.striped.data.dmem);\n"
+    sections += "_bsg_striped_data_start = ADDR(.striped.data.dmem)\n;"
+    sections += "_bsg_dram_t_start_addr = LOADADDR(.text.dram)\n;"
+    sections += "_bsg_dram_d_start_addr = LOADADDR(.text.dram) + SIZEOF(.text.dram)\n;"
+    sections += "_bsg_dram_d_end_addr = LOADADDR(.dram) + SIZEOF(.dram)\n;"
+    sections += "_bsg_dram_end_addr = _bsg_dram_d_end_addr\n;"
+    sections += "_end = _bsg_dram_end_addr\n;"
```

Comment:
Is the function of these variables defined in a google doc somewhere?

Reply:
They are just addresses of boundaries of the memory layout. Start of dmem, end of dmem, start of dram etc. I think verilog loader needs these symbols. May be @tommydcjung can confirm if we still need them...?

Reply:
Are they part of the LLVM Pass?

Reply:
Not sure, but added descriptions for each symbol does.